### DISCUSSION TITLE
Determine the number of output neurons

### DISCUSSION THREAD

**Discussion by davy-blavette** [poloclub/cnn-explainer#36] [General]
Thanks for this work. I am trying to customize the number of output neurons to 7 instead of 10 to test another classification but without success. I have a compatibility problem between the trained model and CNN explainer.
I have customized the number of class on tiny-vgg.py, line 257, NUM_CLASS = 7 and adapted the classLists on config.js but it does not seem sufficient. Can you help me ?

**Comment by xiaohk** [comment]
Hey @davy-blavette, thank you so much for using CNN Explainer!

Currently the visualization is hardcoded to fit the architecture of tiny-vgg (number of layers and number of neurons at each layer). Therefore, to Use CNN Explainer on a different model architecture, you would need to manually modify the visualization code.

You can check out these related issues: #2, #8, #15, #34, #35. Some functions you want to look into are listed [in this comment](https://github.com/poloclub/cnn-explainer/issues/8#issuecomment-624624004).

If you want to change the number of output neurons, you would also need to modify how the flatten layer and logit layer are drawn.

https://github.com/poloclub/cnn-explainer/blob/b9c1759050033850d4245d1a85bb74c475aa377e/src/overview/flatten-draw.js#L990-L998

https://github.com/poloclub/cnn-explainer/blob/b9c1759050033850d4245d1a85bb74c475aa377e/src/overview/flatten-draw.js#L216-L233

**Reply by davy-blavette** [reply]
Hey @xiaohk, Thank you for your answer. The solution was simpler, I just had an error in the generation of my model.json file, with a shape at 1690 for 10 layers instead of 1183 for 7 ... 
However, I had to adapt lines :
replace :
https://github.com/poloclub/cnn-explainer/blob/b9c1759050033850d4245d1a85bb74c475aa377e/src/overview/flatten-draw.js#L859
```python
for (let i = 0; i < nodeCoordinate[curLayerIndex - 1].length; i++) {
```
https://github.com/poloclub/cnn-explainer/blob/b9c1759050033850d4245d1a85bb74c475aa377e/src/overview/flatten-draw.js#L879
```python
textY = nodeCoordinate[curLayerIndex - 1][nodeCoordinate[curLayerIndex - 1].length - 1].y + nodeLength / 2;
```
https://github.com/poloclub/cnn-explainer/blob/b9c1759050033850d4245d1a85bb74c475aa377e/src/overview/flatten-draw.js#L345
```python
let loopFactors = [0, nodeCoordinate[curLayerIndex - 1].length - 1];
```
after add :
https://github.com/poloclub/cnn-explainer/blob/b9c1759050033850d4245d1a85bb74c475aa377e/src/overview/flatten-draw.js#L998-L1002
```python
let maxLayer = nodeCoordinate[curLayerIndex - 1].length - 1;
```
replace :
https://github.com/poloclub/cnn-explainer/blob/b9c1759050033850d4245d1a85bb74c475aa377e/src/overview/flatten-draw.js#L1096
```python
let bottomY = nodeCoordinate[curLayerIndex - 1][maxLayer].y + nodeLength -
```
https://github.com/poloclub/cnn-explainer/blob/b9c1759050033850d4245d1a85bb74c475aa377e/src/overview/flatten-draw.js#L345
```python
let loopFactors = [0, maxLayer];
```
https://github.com/poloclub/cnn-explainer/blob/b9c1759050033850d4245d1a85bb74c475aa377e/src/overview/overview-draw.js#L649-L650
```python
svgPaddings.top + vSpaceAroundGap * (classLists.length) + vSpaceAroundGap +
nodeLength * classLists.length
```


It's ok it works perfectly. Good job !

**Comment by xiaohk** [comment]
Hooray! That's nice @davy-blavette!

Thanks for using the GitHub discussion feature. I believe there will be users having similar questions in the future. I will point them to your thread. (๑•̀ㅂ•́)و✧

**Comment by davy-blavette** [comment]
Hey @xiaohk, so I have integrate cnnExplainer into my datatrainx project. I adapted it to read my own model on a FER2013 base on a single gray channel. You can see it in action on https://datatrainx.akairnet.fr/ via the menu, choose cnnExplainer. The server is on a home network so bad...I tried as much as possible to quote you, but if I've forgotten something, don't hesitate to tell me. Thank you for your work.
![image](https://user-images.githubusercontent.com/3437490/172394497-ef859a41-631d-4a20-a8fc-1defea8d907b.png)

**Comment by xiaohk** [comment]
Hello @davy-blavette! Wow, your demo looks great!! It seems the simple model can classify emotions fairly well.

I really like this figure too:

![image](https://user-images.githubusercontent.com/15007159/172771958-c8f1a5f5-b4f6-4aa5-b891-64ec8d61a8d5.png)

**Reply by davy-blavette** [reply]
Your project @xiaohk, helped me a lot, I bequeath my image rights :joy:
